src/Interpreters/ReplaceQueryParameterVisitor.cpp

```diff
@@ -77,25 +77,26 @@ void ReplaceQueryParameterVisitor::visitQueryParameter(ASTPtr & ast)
 
 void ReplaceQueryParameterVisitor::visitIdentifier(ASTPtr & ast)
 {
-    auto & ast_identifier = ast->as<ASTIdentifier &>();
-    if (ast_identifier.children.empty())
+    auto ast_identifier = dynamic_pointer_cast<ASTIdentifier>(ast);
+    if (ast_identifier->children.empty())
         return;
 
-    auto & name_parts = ast_identifier.name_parts;
+    auto & name_parts = ast_identifier->name_parts;
     for (size_t i = 0, j = 0, size = name_parts.size(); i < size; ++i)
     {
         if (name_parts[i].empty())
         {
-            const auto & ast_param = ast_identifier.children[j++]->as<ASTQueryParameter &>();
+            const auto & ast_param = ast_identifier->children[j++]->as<ASTQueryParameter &>();
             name_parts[i] = getParamValue(ast_param.name);
         }
     }
 
-    if (!ast_identifier.semantic->special && name_parts.size() >= 2)
-        ast_identifier.semantic->table = ast_identifier.name_parts.end()[-2];
+    /// FIXME: what should this mean?
```

Comment:
I don't know too. Maybe we can delete it and see what will happen in tests?

Reply:
I suggest to leave it for the next iteration of refactoring.